browser/src/statechart.js

```diff
@@ -0,0 +1,615 @@
+          actions: 'watershed'
+        }
+      }
+    }
+  },
+  on: {
+    TOGGLERGB: {
+      target: 'paint',
+      in: '#deepcellLabel.rgb.oneChannel'
+    },
+  }
+};
+
+/**
+ * Handles which current tool for mouse behavior.
+ */
+const toolbarState = {
+  initial: 'paint',
+  states: {
+    flood: floodState,
+    trim: trimState,
+    erodeDilate: erodeDilateState,
+    autofit: autofitState,
+    paint: paintState,
+    threshold: thresholdState,
+    watershed: watershedState,
+    hist: { type: 'history' }
+  },
+  on: {
+    LOADING: 'loading',
+    'keydown.b': '.paint',
+    'keydown.g': '.flood',
+    'keydown.k': '.trim',
+    'keydown.q': '.erodeDilate',
+    // tools not available in RGB mode
+    'keydown.t': {
+      target: '.threshold',
+      in: '#deepcellLabel.rgb.oneChannel'
+    },
+    'keydown.m': {
+      target: '.autofit',
+      in: '#deepcellLabel.rgb.oneChannel'
+    },
+    'keydown.w': {
+      target: '.watershed',
+      in: '#deepcellLabel.rgb.oneChannel'
+    },
+  }
+};
+
+/**
+ * Handles tool behavior with a loading state.
+ */
+const mouseState = {
+  initial: 'toolbar',
+  states: {
+    toolbar: toolbarState,
+    loading: {
+      on: { LOADED: 'toolbar.hist' }
+    }
+  },
+  on: {
+    EDIT: { actions: forwardTo('backend') },
+    UNDO: { actions: forwardTo('backend') },
+    REDO: { actions: forwardTo('backend') },
+  }
+};
+
+/**
+ * Handles adjustments to the images displayed on the canvas.
+ */
+const adjusterState = {
+  on: {
+    SCROLLCONTRAST: { actions: 'changeContrast' },
+    SCROLLBRIGHTNESS: { actions: 'changeBrightness' },
+    'keydown.h': { actions: 'toggleHighlight' },
+    'keydown.0': { actions: 'resetBrightnessContrast' },
+    'keydown.i': { actions: 'toggleInvert' }
+  }
+};
+
+/**
+ * Handles moving around the canvas.
+ */
+const canvasState = {
+  initial: 'idle',
+  states: {
+    idle: {
+      on: { 'keydown.space': 'pan' },
+    },
+    pan: {
+      on: {
+        'keyup.space': 'idle',
+        mousemove: { actions: ['pan', 'updateMousePos'] }
+      }
+    }
+  },
+  on: {
+    'keydown.-': { actions: send({ type: 'ZOOM', change: -1 }) },
+    'keydown.=': { actions: send({ type: 'ZOOM', change: 1 }) },
+    ZOOM: { actions: 'zoom' },
+    mousemove: { actions: 'updateMousePos' }
+  }
+};
+
+const brushState = {
+  on: {
+    'keydown.up': { actions: 'incrementBrushSize' },
+    'keydown.down': { actions: 'decrementBrushSize' }
+  }
+};
+
+const displayState = {
+  initial: 'overlay',
+  states: {
+    overlay: {
+      on: {
+        'keydown.z': {
+          target: 'labels',
+          actions: 'updateCanvas'
+        }
+      }
+    },
+    labels: {
+      on: {
+        'keydown.z': {
+          target: 'raw',
+          actions: 'updateCanvas'
+        }
+      }
+    },
+    raw: {
+      on: {
+        'keydown.z': {
+          target: 'overlay',
+          actions: 'updateCanvas'
+        }
+      }
+    }
+  }
+}
+
+const frameState = {
```

Comment:
<img width="533" alt="Screen Shot 2021-02-16 at 2 19 40 PM" src="https://user-images.githubusercontent.com/40068392/108131021-80369280-7065-11eb-8fdc-2f41967d8fea.png">

The frame state lets us to request new images by changing the frame, feature or channel. It includes a loading substate that we transition in/out of on LOADING/LOADED events so we don't request another image while waiting for one.